core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+                if index_id > 0 {
+                    // For non-primary indices we need to first retrieve the primary index key
+                    // through our secondary indices, and then call `.get` again on that.
+                    values
+                        .map(|(_key, value)| {
+                            let raw_row = db.get(&value)
+                                .unwrap()
+                                .expect("secondary index pointed to missing primary key value");
+                            let row: Vec<DataType> = bincode::deserialize(&*raw_row).unwrap();
+                            Row(Rc::new(row))
+                        })
+                        .collect()
+                } else {
+                    values
+                        .map(|(_key, value)| {
+                            let row: Vec<DataType> = bincode::deserialize(&*value).unwrap();
+                            Row(Rc::new(row))
+                        })
+                        .collect()
+                }
+            }
+        };
+
+        LookupResult::Some(Cow::Owned(data))
+    }
+
+    fn add_key(&mut self, columns: &[usize], partial: Option<Vec<Tag>>) {
+        assert!(partial.is_none(), "Bases can't be partial");
+        let existing = self.indices
+            .iter()
+            .any(|index| &index.columns[..] == columns);
+        let is_pk = match self.primary_key {
+            Some(ref pk) => &pk[..] == columns,
+            _ => false,
+        };
+
+        if existing || is_pk {
+            return;
+        }
+
+        let cols = Vec::from(columns);
+        let index_id = self.indices.len() as u32;
+        let mut seq = 0;
+        let mut batch = None;
+
+        // Build the new index for existing values:
+        self.all_rows().for_each(|(ref pk, ref value)| {
+            seq += 1;
+            let row: Vec<DataType> = bincode::deserialize(&value).unwrap();
+            let index_key = columns.iter().map(|i| &row[*i]).collect::<Vec<_>>();
+            let key = self.serialize_key(index_id, &index_key, seq);
+            let b = batch.get_or_insert_with(|| WriteBatch::default());
+            b.put(&key, &pk).unwrap();
+        });
+
+        if let Some(b) = batch {
+            self.db.as_ref().unwrap().write(b).unwrap();
+        }
+
+        self.indices.push(PersistentIndex { columns: cols, seq });
+        self.persist_meta();
+    }
+
+    fn keys(&self) -> Vec<Vec<usize>> {
+        self.indices
+            .iter()
+            .map(|index| index.columns.clone())
+            .collect()
+    }
+
+    fn cloned_records(&self) -> Vec<Vec<DataType>> {
+        self.all_rows()
+            .map(|(_, ref value)| bincode::deserialize(&value).unwrap())
+            .collect()
+    }
+
+    fn rows(&self) -> usize {
+        self.all_rows().count()
+    }
+
+    fn clear(&mut self) {
+        // Would potentially be faster to just drop self.db and call DB::Destroy:
```

Comment:
Yeah, that sounds better. When is this method every called though?

Reply:
The only caller was the `unseed()` function in `MockGraph` so I removed it. Less code 👍